src/server.c

```diff
@@ -144,8 +144,13 @@ void serverLogRaw(int level, const char *msg) {
         } else {
             role_char = (server.masterhost ? 'S':'M'); /* Slave or Master. */
         }
-        fprintf(fp,"%d:%c %s %c %s\n",
-            (int)getpid(),role_char, buf,c[level],msg);
+        if(server.logfmt) {
+            fprintf(fp,"pid=%d role_char=%c timestamp=\"%s\" level=%c message=\"%s\"\n",
+                (int)getpid(),role_char, buf,c[level],msg);
```

Comment:
I suggest `role=%c` (instead of `role_char=%c`). Is one char clear enough? If not, we can consider using role=master, role=sentinel, etc. but it makes the log longer. Maybe one char is clear enough?

For the level, maybe it would be more normal to use the words `warning`, `notice`, `info`, `debug` instead of the symbols?

Do we need to check if `msg` contains double quotes and linebreaks and escape those?

Reply:
One more thing: I think the timestamp should be in ISO format, like `2024-01-11T10:34:38Z` (always in UTC or with another timezone, i'm not sure what's best...)

Reply:
Thanks for your review!
I suppose the questions are for @madolson. I'll wait for his reply.

Reply:
@azuredream The questions are for you. :) What do you think?

Reply:
I actually have a slightly different opinion, which that I think the str format for timestamps should be configurable.

Reply:
As far as I can tell, this is the only case we don't escape: https://github.com/redis/redis/blob/unstable/src/debug.c#L1072. Not sure how we will prevent future usage, also their could be modules that use `"`" in their messages.

Reply:
`"` can also be part of %s arguments if we log key names, user names, etc. We also need to decide what to do with multiline log entries like the redis logo.

Reply:
> " can also be part of %s arguments if we log key names, user names, etc.

We should never be printing out any user provided data into the logs, AFAIK. If we are I would like to go check into that, I would like to remove any chance of accidentally leaking user data into logs.

> We also need to decide what to do with multiline log entries like the redis logo.

I think the right answer is just turning the redis logo off then. I don't think there should be any other multi-line log statements. I would be in favor of adding a `debugServerAssert` validating as much as well.

Reply:
Maybe we can ask people in #12918 about their use scenarios.
IMHO, logfmt is not for human readers, one char is simpler for consumers. Both role and verbose level.
Another concern is consistency among different formats.

Reply:
I assume that configurable timestamp applies for all the log formats.
Some formats for your consideration:

ISO 8601 Format:   2024-01-11T12:45:30+02:00 or 2024-01-11T10:45:30Z
RFC 2822 Format:  Fri, 11 Jan 2024 12:45:30 +0200
Unix Timestamp:   1388534400
RFC 3339 Format:  2024-01-11T12:45:30.05Z or 2024-01-11T14:45:30.05+02:00
Custom Formats:   2024-01-11 12:45:30 GMT+2
Redis Default:        27 Oct 2024 02:30:00.000

Reply:
@madolson the log in debug.c you linked prints the argv, which can be a key name.

    serverLog(LL_WARNING,"client->argv[%d] = \"%s\" (refcount: %d)",
            j, arg, c->argv[j]->refcount);

Reply:
> @madolson the log in debug.c you linked prints the argv, which can be a key name.

Correct, it's on my list of stuff I would like to remove. It was vaguely on my radar as one of the things to fix with https://github.com/redis/redis/pulls?q=is%3Apr+is%3Aopen+redact.

Reply:
@madolson Ah, so do we need to get #11747 merged before we can proceed with this and avoid escaping stuff in message? (I don't like that PRs are blocked by other PRs. Maybe we can just change `"` to `'` in that log entry for now?)

For modules logging, can we simply replace `"` with `'` in-place in moduleLogRaw?

Reply:
> @madolson Ah, so do we need to get https://github.com/redis/redis/pull/11747 merged before we can proceed with this and avoid escaping stuff in message? (I don't like that PRs are blocked by other PRs. Maybe we can just change " to ' in that log entry for now?)

Let's do as you suggested for the moment. Let's also make a debug assertion so we never add the `"` for any logs.

> For modules logging, can we simply replace " with ' in-place in moduleLogRaw?

Like at runtime? I suppose that works, but is also a bit of a breaking change.

Reply:
Yes at runtime. Or replace with `\"`. What else can we do?

Reply:
> Yes at runtime. Or replace with \". What else can we do?

I suppose we probably need to do that now.